### ISSUE TITLE
PCA Hangs on highly correlated data

### ISSUE THREAD

**Issue by SBS-EREHM** [markrogoyski/math-php#414]
[index.php.zip](https://github.com/markrogoyski/math-php/files/6388321/index.php.zip)

// Using PCA to compute a major axis Model II regression, so don't want to center or scale data
// Hangs at "new PCA".  If I set $scale = true, it works, but I don't want to scale data.
// 
// Data
$data = MatrixFactory::create([
      [0.066073, 96.000000],
      [5.407780, 1115.857143],
      [19.440563, 3999.142857],
      [35.582583, 7315.857143],
      [71.602260, 14725.428571],
      [165.725077, 34186.000000],
      [235.426483, 48657.857143],
      [256.868816, 53160.316186]
]);
// Build a principal component analysis model to explore
$center = false;
$scale = false;
$model = new PCA($data, $center, $scale);

$loadings    = $model->getLoadings();     // Matrix of unit eigenvectors of the correlation matrix

$slope = $loadings[0][0] / $loadings[0][1];
echo $slope
// should be approx. 0.0048369

**Comment by SBS-EREHM** [comment]
(base) ERICs-MBP-2:phpPCA ericrehm$ **php --version**
PHP 7.1.23 (cli) (built: Feb 22 2019 22:19:32) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies

**Comment by markrogoyski** [comment]
Hi @SBS-EREHM,

Thank you for your interest in MathPHP.

I was able to reproduce your issue. It goes into a loop which doesn't converge to a diagonal matrix when using the Jacobi method to compute eigenvalues. The math issue here is detecting zero, as floating point math can get you really small values that aren't exactly zero but for things to work need to interpret them as zero. It is oscillating between two really small positive and negative values.

There are two places where this could be fixed. One is in the Jacobi method of the eigenvalues. If it does not converge and oscillates between two values a few times, it could be smart about that and determine it will never converge and end there. Another way to do this is to set the error tolerance on the Matrix and have the error tolerance propagate to all derived data matrices. There is infrastructure to have an error tolerance and many methods use that to approximate zero when doing floating point math. The Matrix has a `->setError()` method for adjusting what the error threshold should be for your particular matrix. The problem is, even if it was set, it is not getting passed on to new matrices that are constructed from the base data. It's a tricky problem that requires some refactoring.

To fix this correctly and permanently will require some refactoring and a release. I think I understand what needs to happen. I'll update this thread when I have an update.

In the meantime, if you are comfortable changing the library code locally in your system, you could try adjusting the `ε` constant in the `\MathPHP\Functions\Support` class to something a little smaller, such as `0.00000001`, and your code will probably work and get a reasonable answer.

**Comment by SBS-EREHM** [comment]
Thanks Mark!  That’s what I suspected.  I can achieve convergence with ε = 0.00000000015;

I’m in the unenviable situation where I need PCA results from three languages: Python (sklearn), JavaScript (ml-pca) and PHP (math-php) to agree on the same input data. Unfortunately, the eigenvectors from math-php don’t as closely match the results (> 3 significant digits) from the JavaScript and Python PCA implementations as I would like.  (Just the first row of eigenvectors are shown.)

Python:

Eigenvalues: [474055727.4,  0.07621130960]
Eigenvectors: [0.004836894105, 00.9999883022]
Slope = evec[0]  / evec[1] = 0.00483695

JavaScript:

Eigenvalues:  [474055727.4, 0.07621118906]
Eigenvectors: [0.004836894104, -0.9999883022]
Slope = evec[0]  / evec[1] = 0.00483695

Php :

Eigenvalues: [474055727.4, 0.07714023733]
Eigenvectors: [0.004838294011, 0.9999882954]
Slope = evec[0]  / evec[1] = 0.00483835

I know this is a tricky issue with respect to how certain calculations (e.g., Jacobi method) are implemented in different PCA approaches, centering, scaling, etc.  FYI, I want to use PCA on unscaled/uncentered data as this is equivalent to a Model II Major Axis linear regression.  Unfortunately, centere & /scaled data corresponds to another approach (Reduced Major Axis regression) that I don’t want to do for my application.

Thoughts?

- eric

On Apr 27, 2021, at 6:51 PM, Mark Rogoyski ***@***.******@***.***>> wrote:



Hi @SBS-EREHM<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_SBS-2DEREHM&d=DwMFaQ&c=9mghv0deYPYDGP-W745IEdQLV1kHpn4XJRvR6xMRXtA&r=oyDiokJpqC12t0ZN_mTYGu2t7Pnrk0fjHVsGJ33nKtw&m=jvvsoWbd5Uv2NTEvH9lEJlEc6giRrV67Qm36mPDzb6w&s=shrWXGOHqHkvDb6e-dHzZXMzQ1WyGvY_annJJ5ulcLc&e=>,

Thank you for your interest in MathPHP.

I was able to reproduce your issue. It goes into a loop which doesn't converge to a diagonal matrix when using the Jacobi method to compute eigenvalues. The math issue here is detecting zero, as floating point math can get you really small values that aren't exactly zero but for things to work need to interpret them as zero. It is oscillating between two really small positive and negative values.

There are two places where this could be fixed. One is in the Jacobi method of the eigenvalues. If it does not converge and oscillates between two values a few times, it could be smart about that and determine it will never converge and end there. Another way to do this is to set the error tolerance on the Matrix and have the error tolerance propagate to all derived data matrices. There is infrastructure to have an error tolerance and many methods use that to approximate zero when doing floating point math. The Matrix has a ->setError() method for adjusting what the error threshold should be for your particular matrix. The problem is, even if it was set, it is not getting passed on to new matrices that are constructed from the base data. It's a tricky problem that requires some refactoring.

To fix this correctly and permanently will require some refactoring and a release. I think I understand what needs to happen. I'll update this thread when I have an update.

In the meantime, if you are comfortable changing the library code locally in your system, you could try adjusting the ε constant in the \MathPHP\Functions\Support class to something a little smaller, such as 0.00000001, and your code will probably work and get a reasonable answer.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_markrogoyski_math-2Dphp_issues_414-23issuecomment-2D828079636&d=DwMFaQ&c=9mghv0deYPYDGP-W745IEdQLV1kHpn4XJRvR6xMRXtA&r=oyDiokJpqC12t0ZN_mTYGu2t7Pnrk0fjHVsGJ33nKtw&m=jvvsoWbd5Uv2NTEvH9lEJlEc6giRrV67Qm36mPDzb6w&s=9fG-cPXpl6LdqotznRVDt2hp2kHmwx_NzJHSDjGQKog&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AQJ4AMLSY3HEMEAIISBLFELTK5SZPANCNFSM43VL3NMA&d=DwMFaQ&c=9mghv0deYPYDGP-W745IEdQLV1kHpn4XJRvR6xMRXtA&r=oyDiokJpqC12t0ZN_mTYGu2t7Pnrk0fjHVsGJ33nKtw&m=jvvsoWbd5Uv2NTEvH9lEJlEc6giRrV67Qm36mPDzb6w&s=zFlaQiwfXmOyc4_-TykFBSM6VsKCArh6MnkL8YUEH7c&e=>.

Please be advised that this email may contain confidential information. If you are not the intended recipient, please notify us by email by replying to the sender and delete this message. The sender disclaims that the content of this email constitutes an offer to enter into, or the acceptance of, any agreement; provided that the foregoing does not invalidate the binding effect of any digital or other electronic reproduction of a manual signature that is included in any attachment.

**Comment by markrogoyski** [comment]
Hi @SBS-EREHM,

Just to get another opinion on this calculation, I plugged in your data into R.

```R
> library(mdatools)
> 
> data = rbind(c(0.066073, 96.000000), c(5.407780, 1115.857143), c(19.440563, 3999.142857), c(35.582583, 7315.857143), c(71.602260, 14725.428571), c(165.725077, 34186.000000), c(235.426483, 48657.857143), c(256.868816, 53160.316186))
> data
           [,1]      [,2]
[1,]   0.066073    96.000
[2,]   5.407780  1115.857
[3,]  19.440563  3999.143
[4,]  35.582583  7315.857
[5,]  71.602260 14725.429
[6,] 165.725077 34186.000
[7,] 235.426483 48657.857
[8,] 256.868816 53160.316
> 
> model = pca(data, center=FALSE, scale=FALSE)

> loadings = model$loadings
> loadings
           Comp 1       Comp 2
[1,] -0.004838294 -0.999988295
[2,] -0.999988295  0.004838294

> slope = loadings[1] / loadings[2]
> slope
[1] 0.004838351
```

The slope R gets looks to be exactly the same that you are getting with MathPHP.

**Comment by markrogoyski** [comment]
Hi @SBS-EREHM,

I looked into Python's sklearn a little, and I think they center the data, but don't scale it, before doing calculations.

Therefore, in MathPHP if you set center to true, you will get an answer that closely resembles your Python calculation.

```php
$center = true;
$scale = false;
$model = new PCA($data, $center, $scale);
```

Doing this I got a value of `0.0048369506856261` for your slope.

**Comment by markrogoyski** [comment]
Hi @SBS-EREHM,

This has been fixed in [v1.11.0](https://github.com/markrogoyski/math-php/releases/tag/v1.11.0) and in the newest release [v2.0.0](https://github.com/markrogoyski/math-php/releases/tag/v2.0.0).

Thank you for reporting this issue. And thanks again for your interest in MathPHP.

Mark
